### PR TITLE
fix: follow dependencies when looking for a struct

### LINKED ISSUE #3385: Re-exports cannot be used as a parameter in the main function
### Aim

I have created this PR within a minimal reproducable example: https://github.com/noir-lang/noir/pull/3384

There are three crates:

- library2
- library
- binary

- Library2 defines a type Foo.
- Library depends on Library and re-exports Foo.
- Binary then depends on Library and uses Foo as a parameter in their main function.

### Expected Behavior

This should compile

### Bug

The application panicked (crashed).
Message:  The Struct was supposed to be defined in a dependency
Location: compiler/noirc_frontend/src/hir/mod.rs:129

### To Reproduce

1.
2.
3.
4.

### Installation Method

None

### Nargo Version

_No response_

### Additional Context

_No response_

### Would you like to submit a PR for this Issue?

No

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by guipublic**
# Description

## Problem\*

Resolves #3385
## Summary\*

When we want to fully qualify a struct, we follow recursively dependencies until we get to the crate defining the struct.

## Documentation\*

Check one:
- [X] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[Exceptional Case]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by kevaundray** [comment]
We don't have regression tests, could you run this on a few examples to see that this did not regress, then open up an issue on how we would test regressions for this?

File: compiler/noirc_frontend/src/hir/mod.rs [hunk 1]
```diff
@@ -119,18 +119,34 @@ impl Context {
...
+                fqn += &name.to_string();
+                fqn += "::";
+            }
+            fqn += &module_path;
+            fqn
         }
     }
 
+    /// Recursively walks down the crate dependency graph from crate_id until we reach requested crate
+    /// returns the path from crate_id to target_crate_id
```
[the remark this answers is not in the document]
Comment (guipublic) [reply compiler/noirc_frontend/src/hir/mod.rs]: Done

**Comment by guipublic** [comment]
> We don't have regression tests, could you run this on a few examples to see that this did not regress, then open up an issue on how we would test regressions for this?

I don't understand, I have added a regression test for this fix. Are you talking about something else?

File: compiler/noirc_frontend/src/hir/mod.rs [hunk 2]
```diff
@@ -119,18 +119,37 @@ impl Context {
...
+                return vec![dep.name.clone()];
+            }
+            let mut fqn = self.find_dependencies(&dep.crate_id, target_crate_id);
+            if !fqn.is_empty() {
+                fqn.insert(0, dep.name.clone());
+                return fqn;
+            }
+        }
+        Vec::new()
+    }
```
Comment (jfecher) [inline compiler/noirc_frontend/src/hir/mod.rs:151]:
```suggestion
    fn find_dependencies(&self, crate_id: &CrateId, target_crate_id: &CrateId) -> Option<Vec<String>> {
        for dep in &self.crate_graph[crate_id].dependencies {
            if &dep.crate_id == target_crate_id {
                return Some(vec![dep.name.clone()]);
            }
            
            if let Some(mut path) = self.find_dependencies(&dep.crate_id, target_crate_id) {
                path.insert(0, dep.name.to_string());
                return path;
            }
        }
        None
    }
```
- I think using an `Option` makes the failure case a bit clearer
- Returning Strings directly lets us use `join` when this function is called above.
- I tried changing to a VecDeque, but that doesn't support `join` unfortunately.

File: compiler/noirc_frontend/src/hir/mod.rs [hunk 3]
```diff
@@ -119,18 +119,37 @@ impl Context {
...
-            format!("{crate_name}::{module_path}")
+            let crates = self.find_dependencies(crate_id, &module_id.krate);
+            assert!(!crates.is_empty(), "The Struct was supposed to be defined in a dependency");
+            let mut fqn = String::new();
+            for name in crates {
+                fqn += &name.to_string();
+                fqn += "::";
+            }
+            fqn += &module_path;
+            fqn
```
Comment (jfecher) [inline compiler/noirc_frontend/src/hir/mod.rs:130]:
```suggestion
            let crates = self.find_dependencies(crate_id, &module_id.krate)
                .expect("The Struct was supposed to be defined in a dependency");
            crates.join("::") + "::" + &module_path
```

Comment (guipublic) [reply compiler/noirc_frontend/src/hir/mod.rs]: Done

File: compiler/noirc_frontend/src/hir/mod.rs (hunk 2 above)
Comment (guipublic) [reply compiler/noirc_frontend/src/hir/mod.rs]: Done

**Comment by guipublic** [comment]
> Yep was referring to it not changing the behaviour of programs that were previously passing (..)

I have created an issue for this: #3421 
What I could also do is modify the logic of this PR to use the previous method and fallback to the new one when it is failing. That way we are sure that previous programs keep the same behaviour.
Let me know if you want me to do this.

**Review by jfecher — APPROVED** [review]
LGTM

**Comment by kevaundray** [comment]
I think just manually checking would have been good to merge -- the importance being that aztec-packages depends on this for doc generation

**Pull request merged**